Add Gateway API GA blog post (#43593)
kubernetes/website: content/en/blog/_posts/2023-10-19-Gateway-API-GA.md

Reviewer: Overall
- please write headings in sentence case
- don't use block quotes unless you are quoting from elsewhere (you can use Docsy [alerts](https://www.docsy.dev/docs/adding-content/shortcodes/#alert) though)
  - use these very sparingly. One or two is enough.
- put the URL for hyperlinks inline, not separately
- don't put versions in backticks, with the exceptions that version strings you see in API URLs (eg `v1beta1`) are OK

Reviewer 2: Thanks @sftim! I've pushed a significant update that should resolve all of these.

> order to be considered supported.

### Getting Started

If you want to try Gateway API features out for yourself but don't know where
to get started you'll want to check out one of the 20+
[implementations](https://gateway-api.sigs.k8s.io/implementations/) which
we list on the Gateway API website.

Gateway API itself is just an API for downstream projects to implement, so take
some time to peruse the implementations list above and find one that seems to
fit your needs, or that you're most familiar with.

> **Note**: with this release we now have basic support for downstream
> implementations to report the results of their Gateway API conformance test
> runs. This means that for some implementations on the above implementations
> page you may see a `Gateway API Conformance vX.X.X` badge present under their
> project header. This indicates that this implementation _is conformant_ for
> the version present in the badge, and is sending back reports of their
> conformance to the upstream project, including details on which specific
> features they support. Click on the badge to see a more detailed report of
> supported features, but note that the reporting system is considered

> > experimental and the structure of these reports is subject to change.

Reviewer 2: I'm going to remove this from the blog post in favor of a link to the getting started guide on Gateway API. I think we should include a statement like this on the Gateway API website though, maybe directly on the implementations page?

Author: sure

---
layout: blog
title: "Gateway API is now Generally Available (GA)"
date: 2023-10-19T10:00:00-08:00
slug: gateway-api-ga
---

***Authors:*** Shane Utt (Kong), Nick Young (Isovalent), Rob Scott (Google)

We (the community developing [Gateway API][gwapi]) are incredibly pleased and
proud to announce _general availability_ of Gateway API for use with Kubernetes
clusters `v1.24` and newer!

## Important Highlights

> Gateway API brings a lot of new networking possibilities to the table in your

Reviewer: (nit)
```suggestion
Kubernetes' Gateway API brings a lot of new networking possibilities to the table in your
```

Kubernetes cluster. In this post we'll cover some of the most important
highlights.

For a complete list of changes for this specific release, please see our
[v1.0.0 changelog entry][chlogv1].

For a complete list of every feature for every release, or to review some of

> the releases leading up to `v1.0.0`, please see our [historical

Reviewer: ```suggestion
the releases leading up to v1.0.0, please see our [historical
```

changelog][clog].

### Version 1 APIs

This marks the **first stable release** of Gateway API, where we are now

> publishing the following APIs as `v1` (Version 1):

Reviewer: ```suggestion
publishing the following APIs as `v1` (version 1):
```

- [GatewayClass][gwc]
- [Gateway][gw]
- [HTTPRoute][hr]

Reaching version 1 with these APIs means we future iterations on them will be

> handled in a backwards compatible manner.

Reviewer: Might want to clarify how closely Gateway will / won't follow https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions and https://kubernetes.io/docs/reference/using-api/deprecation-policy/
(linking to a document on this, covering Gateway's specific policy, is fine).

Reviewer 3: ~we~ or reword to:

```
Reaching version 1 with these APIs will follow the [stable](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions) API policy.
```

Reviewer 2: Good call. I just pushed a significant update that include a link to Gateway API's versioning policy.

Gateway API [CRDs][crds] now include [CEL][cel] validation. This streamlines
the development of validation for our resources, and reduces the overhead of
deploying Gateway API to a cluster as it eliminates the need for our historical
[admission webhook][admw].

> **Note**: For Kubernetes `v1.25` and newer we recommend against deploying the
> webhook, and if you had it running previously you can now uninstall it in
> favor of the validation now provided via CEL.

### Kubernetes Support

Gateway API generally [supports the last 5 stable Kubernetes versions][kvs]. As
such, this release is supported on Kubernetes clusters version `1.24` through
`1.29` at the time of release.

> **Warning**: Related to our migration to [CEL][cel] as mentioned above, we
> don't recommend deploying on `v1.24` clusters if it can be avoided as these
> clusters will still require the admission webhook, and that webhook is
> considered deprecated and support for it will drop shortly after this GA
> release when Kubernetes `v1.30` becomes available. If it's unavoidable to
> deploy on `v1.24`, just note that you'll need to upgrade fairly soon in
> order to be considered supported.

### Getting Started

> If you want to try Gateway API features out for yourself but don't know where
> to get started you'll want to check out one of the 20+ implementations which
> we list on our website:
> 
> https://gateway-api.sigs.k8s.io/implementations/

Reviewer: ```suggestion
If you want to try Gateway API features out for yourself but don't know where
to get started you'll want to check out one of the 20+
[implementations](https://gateway-api.sigs.k8s.io/implementations/) which
we list on the Gateway API website.
```

Any plans to add something to CNCF Landscape (or similar)?

Author: No plans for that currently in fact, maybe it makes sense to do so after GA
/cc @robscott @youngnick

> the version present in the badge, and is sending back reports of their
> conformance to the upstream project, including details on which specific
> features they support. Click on the badge to see a more detailed report of
> supported features, but note that the reporting system is considered
> experimental and the structure of these reports is subject to change.

## Future Plans

General availability is just the beginning of the much larger journey for
Gateway API, and there's plenty of new features and new ideas in flight for the
upcoming releases!

If you opt-in to the [experimental channel][expch] for this release, you'll find
that there are several experimental features available which are being
considered for inclusion as stable in future releases:

- [BackendTLSPolicy][gep1897]
- [HTTPRoute Timeouts][gep1742]
- [Gateway Infrastructure Labels][gep1762]
- [Websockets, HTTP/2 and More!][gep1911]
- [GWCTL, our new Gateway API CLI][gwctl]

We intend to have a blog post following up on this one which will cover these
(and potentially more) experimental and upcoming features where the authors of
the features themselves can provide a greater level of detail.

> In future iterations we also intend to grow the maturity and work towards
> [graduation][grad] of APIs other than `GatewayClass`, `Gateway`, and
> `HTTPRoute`, which are not yet at version 1:

Reviewer: ```suggestion
In future iterations we also intend to grow the maturity and work towards
[graduation][grad] of APIs other than GatewayClass, Gateway, and
HTTPRoute. There are several APIs that are not yet at version 1:
```

- [GRPCRoute][grpc]
- [TCPRoute][tcp]
- [UDPRoute][udp]
- [TLSRoute][tls]
- [ReferenceGrant][refg]

[ReferenceGrant][refg] is a slightly special case as we are considering what

> to do with this resource next: we've come to understand this might be a

Reviewer: ```suggestion
to do with this: we've come to understand this might be a
```

to avoid overloading “resource” (lots of meanings!)

feature applicable to the greater Kubernetes community and we're trying to
navigate potentially moving it somewhere more general to maximize benefit.

For those interested in service mesh, note that we continue to work on making
Gateway API resources available for use in service mesh contexts as part of the

> [GAMMA project][gamma]. Please see our recent blog post ["Gateway API v0.8.0:
> Introducing Service Mesh Support"][gammablog] for more details.

Reviewer: ```suggestion
[GAMMA project][gamma]. Please see our recent blog post [Gateway API v0.8.0:
Introducing Service Mesh Support][gammablog] for more details.
```
